src/canvas.ts

```diff
@@ -0,0 +1,138 @@
+import { Renderer, basePoint } from './utilities/renderer';
+
+type DotProperties = Point;
+type LineProperties = Line;
+type RectangleProperties = Omit<Line, 'end'> & Partial<Pick<Line, 'end'>> & Partial<Size>;
+interface CircleProperties {
+  center: Point;
+  radius: number;
+}
+interface TextProperties {
+  start: Point;
+  text: string;
+  size: number;
+}
+interface ImageProperties {
+  image: CanvasImageSource;
+  start: Point;
+  imageStart: Point;
+  size: Size;
+  ratio?: number;
+}
+
+export class Canvas {
+  private _renderer: Renderer | undefined;
+
+  constructor(protected element: HTMLCanvasElement) {}
+
+  get width(): number {
+    return parseInt(this.element.getAttribute('width') ?? '0');
+  }
+  set width(value: number) {
+    this.element.setAttribute('width', `${value}`);
+  }
+  get height(): number {
+    return parseInt(this.element.getAttribute('height') ?? '0');
+  }
+  set height(value: number) {
+    this.element.setAttribute('height', `${value}`);
+  }
+  get size(): Size {
+    const { width, height }: Canvas = this;
+    return { width, height };
+  }
+
+  get renderer(): Renderer {
+    if (!this._renderer) {
+      this._renderer = new Renderer(basePoint, basePoint, basePoint, basePoint, 0);
+    }
+    return this._renderer;
+  }
+  set renderer(value: Renderer) {
+    this._renderer = value;
+  }
+
+  private get context(): CanvasRenderingContext2D {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const context: CanvasRenderingContext2D = this.element.getContext('2d')!;
+    context.save();
+    context.beginPath();
+    return context;
+  }
+
+  private color(context: CanvasRenderingContext2D, color: string) {
+    context.fillStyle = color;
+    context.strokeStyle = color;
+    return this.getDrawBuilder(context);
+  }
+  private lineWidth(context: CanvasRenderingContext2D, lineWidth: number) {
```

Comment:
- The class member `_renderer` need not be initialized upfront to `undefined`.
- Code in the `dot`, `line`, `rectangle`, `circle`, `text`, and `image` function should be wrapped with `beginPath()` and `closePath()`.
- The `getDrawBuilder` function unnecessarily repeats the `fill` implementation. It should be removed.
- In the `color` and `lineWidth` functions, the call to `this.getDrawBuilder` is redundant.
- The `text` and `image` functions don't need to restore the context (`restore()`).
- Parameter `width` and `height` in interface `RectangleProperties` should be required.
- The first sentence in `TextProperties`'s interface description should be capitalized for consistency.
- Parameter `ratio` in interface `ImageProperties` should be optional.
- Padding of an empty line should be removed after the `break` statement in `fill` and `stroke` functions.

{ "start": 1, "end": 68 }